### PR TITLE
chore: Post-C4 Re-review Changes (SC-2103)

### PR DESCRIPTION
**Pull request by deluca-mike**
# Description

Post-C4 Re-review Changes and Custodian Allowance considerations

# Integrations Checklist

- [X] Have any function signatures changed? If yes, outline below.
- [X] Have any features changed or been added? If yes, outline below.
- [X] Have any events changed or been added? If yes, outline below.
- [X] Has all documentation been updated?

# Changelog

- Renamed incorrectly named `FDT` contract and `IFDT` interface to `LoanFDT` and `ILoanFDT` respectively.
- Emitted a missing `BalanceUpdated` event in `Loan.withdrawFunds`.
- Put reused `Pool` code into a new internal `_canWithdraw` function.
- Slight reduction in gas by skipping the need for a `_globals` function in `PoolLib`, as it was only used once.
- Slight reduction in gas by making `PoolLib.bdiv` internal.
- Fixed last few NatSpec indentations issues
- Fixed typos in test comments
- Fixed incorrect NatSpec title for `StakeLockerFDT`
- `PoolDelegateAllowed` event renamed to `PoolDelegateSet`
- `_checkPercentageRange` now accepts the upper bound as an argument
- `setInvestorFee` and `setTreasuryFee` reuse new `_checkFee`, which itself uses updated `_checkPercentageRange`
- Gas savings by using `msg.sender` instead of `pendingGovernor` when setting `governor`
- `TotalCustodyAllowanceUpdated` emissions in `Pool` and `StakeLocker`
- Added more NatSpec for event emissions

### CODE REVIEW

File: contracts/Loan.sol [hunk 1]
```diff
@@ -436,10 +439,12 @@ contract Loan is FDT, Pausable {
 
     /**
         @dev Withdraws all available funds earned through FDT for a token holder.
+        @dev It emits a `BalanceUpdated` event.
     */
     function withdrawFunds() public override {
         _whenProtocolNotPaused();
         super.withdrawFunds();
+        emit BalanceUpdated(address(this), address(fundsToken), fundsToken.balanceOf(address(this)));
```
Comment (deluca-mike) [inline contracts/Loan.sol:447]: This was missing.

File: contracts/MapleGlobals.sol [hunk 2]
```diff
@@ -63,7 +63,7 @@ contract MapleGlobals {
     event               GlobalsAddressSet(bytes32 indexed which, address addr);
     event                  ProtocolPaused(bool pause);
     event                  GlobalAdminSet(address newGlobalAdmin);
-    event             PoolDelegateAllowed(address delegate, bool valid);
+    event                 PoolDelegateSet(address delegate, bool valid);
```
Comment (deluca-mike) [inline contracts/MapleGlobals.sol:66]: [See](https://github.com/maple-labs/maple-core/pull/386#discussion_r623042790)

File: contracts/MapleGlobals.sol [hunk 3]
```diff
@@ -152,7 +152,7 @@ contract MapleGlobals {
         @param newMaxSlippage New max slippage percentage (in basis points)
     */
     function setMaxSwapSlippage(uint256 newMaxSlippage) external isGovernor {
-        _checkPercentageRange(newMaxSlippage);
+        _checkPercentageRange(newMaxSlippage, 10_000);
```
Comment (deluca-mike) [inline contracts/MapleGlobals.sol:155]: [See](https://github.com/maple-labs/maple-core/pull/386#discussion_r623047223)

File: contracts/MapleGlobals.sol [hunk 4]
```diff
@@ -281,8 +281,7 @@ contract MapleGlobals {
         @param _fee The fee, e.g., 50 = 0.50%.
     */
     function setInvestorFee(uint256 _fee) external isGovernor {
-        _checkPercentageRange(_fee);
-        require((_fee + treasuryFee) <= 10_000, "MG:INVALID_INVESTOR_FEE");
+        _checkFee(_fee, _fee + treasuryFee);
```
Comment (deluca-mike) [inline contracts/MapleGlobals.sol:284]: [See](https://github.com/maple-labs/maple-core/pull/386#discussion_r623047223)

File: contracts/MapleGlobals.sol [hunk 5]
```diff
@@ -293,8 +292,7 @@ contract MapleGlobals {
         @param _fee The fee, e.g., 50 = 0.50%.
     */
     function setTreasuryFee(uint256 _fee) external isGovernor {
-        _checkPercentageRange(_fee);
-        require((_fee + investorFee) <= 10_000, "MG:INVALID_TREASURY_FEE");
+        _checkFee(_fee, _fee + investorFee);
```
Comment (deluca-mike) [inline contracts/MapleGlobals.sol:295]: Ditto

File: contracts/MapleGlobals.sol [hunk 6]
```diff
@@ -384,7 +382,7 @@ contract MapleGlobals {
     */
     function acceptGovernor() external {
         require(msg.sender == pendingGovernor, "MG:NOT_PENDING_GOV");
-        governor = pendingGovernor;
+        governor        = msg.sender;
```
Comment (deluca-mike) [inline contracts/MapleGlobals.sol:385]: [See](https://github.com/maple-labs/maple-core/pull/386#discussion_r623048184)

File: contracts/MapleGlobals.sol [hunk 7]
```diff
@@ -439,9 +437,17 @@ contract MapleGlobals {
...
+        _checkPercentageRange(_fee, 10_000);
+        _checkPercentageRange(_fee, _totalFees);
+    }  
+
+    /**
+        @dev Checks that percentage is less than a specified max.
+    */
+    function _checkPercentageRange(uint256 percentage, uint256 max) internal pure {
+        require(percentage <= max, "MG:PCT_OOB");
     }
```
Comment (deluca-mike) [inline contracts/MapleGlobals.sol:452]: [See](https://github.com/maple-labs/maple-core/pull/386#discussion_r623047223)

File: contracts/Pool.sol [hunk 8]
```diff
@@ -58,20 +58,21 @@ contract Pool is PoolFDT {
...
+    event              LPStatusChanged(address indexed user, bool status);
+    event              LiquidityCapSet(uint256 newLiquidityCap);
+    event              LockupPeriodSet(uint256 newLockupPeriod);
+    event                StakingFeeSet(uint256 newStakingFee);
+    event             PoolStateChanged(State state);
+    event                     Cooldown(address indexed lp, uint256 cooldown);
+    event           PoolOpenedToPublic(bool isOpen);
+    event                 PoolAdminSet(address poolAdmin, bool allowed);
+    event           DepositDateUpdated(address indexed lp, uint256 depositDate);
+    event TotalCustodyAllowanceUpdated(address indexed tokenHolder, uint256 newTotalAllowance);
```
Comment (deluca-mike) [inline contracts/Pool.sol:75]: [See](https://github.com/maple-labs/maple-core/pull/386#discussion_r623121838)

File: contracts/Pool.sol [hunk 9]
```diff
@@ -406,19 +409,28 @@ contract Pool is PoolFDT {
...
+    /**
+        @dev   Checks that the user can withdraw an amount.
+        @param user The address of the user.
+        @param wad  The amount to withdraw.
+    */
+    function _canWithdraw(address user, uint256 wad) internal view {
+        require(depositDate[user].add(lockupPeriod) <= block.timestamp,  "P:FUNDS_LOCKED");     // Restrict transfer during lockup period
+        require(balanceOf(user).sub(wad) >= totalCustodyAllowance[user], "P:INSUF_TRANS_BAL");  // User can only withdraw tokens that aren't custodied
+    }
+
```
Comment (deluca-mike) [inline contracts/Pool.sol:421]: Reused by `withdraw` and `_transfer`.

File: contracts/Pool.sol [hunk 10]
```diff
@@ -58,20 +58,21 @@ contract Pool is PoolFDT {
...
-    event         LiquidityCapSet(uint256 newLiquidityCap);
-    event         LockupPeriodSet(uint256 newLockupPeriod);
-    event           StakingFeeSet(uint256 newStakingFee);
-    event        PoolStateChanged(State state);
-    event                Cooldown(address indexed lp, uint256 cooldown);
-    event      PoolOpenedToPublic(bool isOpen);
-    event            PoolAdminSet(address poolAdmin, bool allowed);
-    event      DepositDateUpdated(address indexed lp, uint256 depositDate);
+    event                   LoanFunded(address indexed loan, address debtLocker, uint256 amountFunded);
+    event                        Claim(address indexed loan, uint256 interest, uint256 principal, uint256 fee, uint256 stakeLockerPortion, uint256 poolDelegatePortion);
```
Comment (deluca-mike) [inline contracts/Pool.sol:62]: [See](https://github.com/maple-labs/maple-core/pull/386#discussion_r623142789)

File: contracts/StakeLocker.sol [hunk 11]
```diff
@@ -34,16 +34,17 @@ contract StakeLocker is StakeLockerFDT, Pausable {
...
+    event               BalanceUpdated(address indexed who, address indexed token, uint256 balance);
+    event             AllowListUpdated(address indexed staker, bool status);
+    event             StakeDateUpdated(address indexed staker, uint256 stakeDate);
+    event          LockupPeriodUpdated(uint256 lockupPeriod);
+    event                     Cooldown(address indexed staker, uint256 cooldown);
+    event                        Stake(uint256 amount, address indexed staker);
+    event                      Unstake(uint256 amount, address indexed staker);
+    event              CustodyTransfer(address indexed custodian, address indexed from, address indexed to, uint256 amount);
+    event      CustodyAllowanceChanged(address indexed tokenHolder, address indexed custodian, uint256 oldAllowance, uint256 newAllowance);
+    event TotalCustodyAllowanceUpdated(address indexed tokenHolder, uint256 newTotalAllowance);
```
Comment (deluca-mike) [inline contracts/StakeLocker.sol:47]: [See](https://github.com/maple-labs/maple-core/pull/386#discussion_r623145968)

File: contracts/Loan.sol [hunk 12]
```diff
@@ -18,13 +18,13 @@ import "./interfaces/IUniswapRouter.sol";
 import "./library/Util.sol";
 import "./library/LoanLib.sol";
 
-import "./token/FDT.sol";
+import "./token/LoanFDT.sol";
```
Comment (deluca-mike) [inline contracts/Loan.sol:21]: Was incorrectly named as a generic FDT, when it was specifically the LoanFDT.

File: contracts/library/PoolLib.sol [hunk 13]
```diff
@@ -75,7 +75,7 @@ library PoolLib {
         address dlFactory,
         uint256 amt
     ) external {
-        IMapleGlobals globals = _globals(superFactory);
+        IMapleGlobals globals = IMapleGlobals(ILoanFactory(superFactory).globals());
```
Comment (deluca-mike) [inline contracts/library/PoolLib.sol:78]: `_globals` was only used once, here.

File: contracts/library/PoolLib.sol [hunk 14]
```diff
@@ -260,7 +260,7 @@ library PoolLib {
     /**
         @dev Official balancer pool bdiv() function, does synthetic float with 10^-18 precision
     */
-    function bdiv(uint256 a, uint256 b) public pure returns (uint256) {
+    function _bdiv(uint256 a, uint256 b) internal pure returns (uint256) {
```
Comment (deluca-mike) [inline contracts/library/PoolLib.sol:263]: Only used internally.

File: contracts/Pool.sol [hunk 15]
```diff
@@ -406,19 +409,28 @@ contract Pool is PoolFDT {
         emit Cooldown(msg.sender, uint256(0));
     }
 
+    /**
+        @dev   Checks that the account can withdraw an amount.
+        @param account The address of the account.
+        @param wad     The amount to withdraw.
+    */
+    function _canWithdraw(address account, uint256 wad) internal view {
+        require(depositDate[account].add(lockupPeriod) <= block.timestamp,     "P:FUNDS_LOCKED");     // Restrict transfer during lockup period
```
Comment (lucas-manuel) [inline contracts/Pool.sol:418]: Should update comment from transfer => withdraw

File: contracts/Pool.sol [hunk 16]
```diff
@@ -441,17 +453,16 @@ contract Pool is PoolFDT {
 
         (uint256 lpCooldownPeriod, uint256 lpWithdrawWindow) = _globals(superFactory).getLpCooldownParams();
 
-        require(depositDate[from].add(lockupPeriod) <= block.timestamp,                         "P:FUNDS_LOCKED");            // Restrict transfer during lockup period
-        require(balanceOf(from).sub(wad) >= totalCustodyAllowance[from],                        "P:INSUF_TRANSFERABLE_BAL");  // User can only transfer tokens that aren't custodied
-        require(block.timestamp > (withdrawCooldown[to] + lpCooldownPeriod + lpWithdrawWindow), "P:TO_NOT_ALLOWED");          // Recipient must not be currently withdrawing
-        require(recognizableLossesOf(from) == uint256(0),                                       "P:RECOG_LOSSES");            // If an LP has unrecognized losses, they must recognize losses through withdraw
+        _canWithdraw(from, wad);
+        require(block.timestamp > (withdrawCooldown[to] + lpCooldownPeriod + lpWithdrawWindow), "P:TO_NOT_ALLOWED");   // Recipient must not be currently withdrawing
```
Comment (lucas-manuel) [inline contracts/Pool.sol:457]: Extra space between semicolon and //

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Small changes

File: contracts/MapleGlobals.sol [hunk 17]
```diff
@@ -439,7 +437,7 @@ contract MapleGlobals {
     /************************/
 
     /**
-        @dev Checks that percentage is less than 100%.
+        @dev Checks that percentage is less than a specified max.
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply contracts/MapleGlobals.sol]: That was a mistake. I was positive I had reverted this change...

[the remark this answers is not in the document]
Comment (deluca-mike) [reply contracts/MapleGlobals.sol]: Ah, I did. Your comment is outdated.

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**